src/rdf4cpp/rdf/Dataset.cpp

```diff
@@ -10,25 +10,30 @@ Dataset::Dataset(Dataset::NodeStorage node_storage)
     : dataset_storage(Dataset::DatasetStorage::new_instance<storage::tuple::DefaultDatasetBackend>(std::move(node_storage))) {}
 
 void Dataset::add(const Quad &quad) {
-    dataset_storage.add(quad.to_node_storage(backend().node_storage()));
+    auto ns = backend().node_storage().upgrade();
+    dataset_storage.add(quad.to_node_storage(ns));
```

Comment:
I guess this would get a bit more expensive that way.

Reply:
reverted